freqtrade/optimize/hyperopt.py

```diff
@@ -409,6 +409,52 @@ def prepare_hyperopt_data(self) -> None:
         # Store non-trimmed data - will be trimmed after signal generation.
         dump(preprocessed, self.data_pickle_file)
 
+    def get_asked_points(self, n_points: int) -> Tuple[List[List[Any]], List[bool]]:
+        '''
+        Enforce points returned from `self.opt.ask` have not been already evaluated
+
+        Steps:
+        1. Try to get points using `self.opt.ask` first
+        2. Discard the points that have already been evaluated
+        3. Retry using `self.opt.ask` up to 3 times
+        4. If still some points are missing in respect to `n_points`, random sample some points
+        5. Repeat until at least `n_points` points in the `asked_non_tried` list
+        6. Return a list with length truncated at `n_points`
+        '''
+        def unique_list(a_list):
+            seen = []
+            for x in a_list:
+                key = repr(x)
+                if key not in seen:
+                    seen.append(key)
+            return [eval(x) for x in seen]
```

Comment:
I think we should try to avoid using `eval` if possible (and i think it's possible above, and don't understand why it's made like this).
it's a quite dangerous method (it executes whatever you pass in).

in my understanding of the code - we show the representation (based on the assumption that it will come out as `"float(5)"` or something in a similar format).

We'll then append it to the seen list - and before returning, eval it (basically executing the above string - so `float(5)` turns into `5.0`).

This is unfortunately bound to fail / raise errors since
* you can't rely on all types to implement `__repr__` as `Type(value)`
* even if they implement it this way - nothing guarantees that this will give you the correct value (maybe the correct initialization would be `Type(value, somethingelse)`

I'm not sure why you convert it to string (and back) - it'll create the same object twice (performance problem) - and potentially result in value changes (if the `__repr__` limits precision shown, for example).

Reply:
Makes sense. I used that because the list is a list of lists that is unhashable if you try to transform it into a set. I've found another way to achieve the same with a simple for loop (just pushed).

Reply:
Makes sense. I used that because the list is a list of lists that is unhashable if you try to transform it into a set. I've found another way to achieve the same with a simple for loop (just pushed).